상품 상세페이지 구현 PR

silvertae opened the pull request:
## Description
- 판매자/구매자 상품 상세 페이지로 사용할 ItemDetails 컴포넌트 구현
- 상세 페이지 관련 api axios요청, 리액트 쿼리, mock data 추가

## Key changes
### 브라우저 환경 (판매자 페이지 기준)
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/2a2527af-ba06-48de-aa48-15fe778ab91d

### 구매자 페이지
![스크린샷 2023-09-14 오후 12 21 11](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/f3fa4ba9-c3be-4b19-abe1-d8f6c46c695c)

### 모바일
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/2094f6e9-7a7a-4690-aebf-440e24f0d4d9

- `ItemDetails` 라는 이름의 컴포넌트로 상품 상세 페이지를 구현하였습니다. 우선 리액트 라우터를 사용해 `/items/:itemId`라는 경로를 가진 별도의 페이지로 navigate되는 형식으로 구현하였습니다. 
  - 이유는 리액트를 라우터를 적용했는데 뭔가 너무 한 페이지에서 패널로 다 처리하고 있는 것 것 같아서(?) 사실 다양한 방식을 시도해 보고싶었던 것도 있네요. 추후 페이지 이동을 어떤 식으로 할지 제이랑 의논해보고 패널로 바꿔도 될 것 같습니다
  - 만약 패널로 안하고 페이지 라우팅 시 애니메이션을 적용한다고 하면 react transition group을 활용하는 방법도 있는 것 같더라구요. 흥미로워서 첨부해봅니다. https://devnm.tistory.com/10 
- 상품 이미지가 표시되는 영역은 `ImageSlider` 라는 이름의 별도 컴포넌트로 구현하였습니다. 브라우저에서는 클릭-드래그해서, 모바일에서는 터치-드래그 해서 이미지를 넘겨볼 수 있습니다.
- 공통 컴포넌트들 중에 약간 수정된게 있는데 주로 props에 { ...rest }를 추가한 부분입니다. ItemDetails 구현시 기존 공통 컴포넌트에서 스타일이 다른 부분이 있어서 styled-components로 확장시켜서 사용할 때 props에 `className`이 필수더라구요.
- Dropdown, ProductItem에서 기획서랑 다르게 여백이 좀 이상한 부분이 있어 추가로 수정했습니다.
- item 관련 요청들이 너무 한 파일에 몰려있는 것 같아 필요한 코드 찾는게 번거롭다고 생각돼서 우선 itemDetails 관련 api 요청들을 별도 파일들로 분리해둔 상태입니다. 
  - 기존 각 파일마다 관리하던 queryKey들도 endPoint 처럼 별도 파일에 만들어두고 쓰면 좋을 것 같아 수정해보았습니다. 
  - fetcher 함수들도 많아진 것 같아 api 폴더에 fetchers 라는 폴더를 따로 두었는데 괜찮은지 확인 부탁드립니다.
  - 
## To reviewers
생각보다 뭔가 이것저것 한 것 치고는 아직 고쳐야 할 부분이 많이 보이네요... 
특히 이미지 슬라이더 같은 경우는 scrollLeft를 즉각적으로 변경시키는 식으로 구현되어 있다보니 부드럽게 넘어가질 않아서 더 공부해서 수정해 볼 예정입니다. 그리고 모바일 환경에서는 이미지 넘길 때 좀 버그가 있더라구요;; 이 점 감안하고 봐주시길... 😅

그리고 아직 기획서 요구사항 중 처음에 헤더가 투명하게 보이다가 스크롤을 하면 배경색이 생기면서 구분선이 생기는 기능은 아직 구현되지 않은 상태입니다!

## Issue
- #64

silvertae committed:
feat #64: 상품 상세 페이지 컴포넌트 구조 잡기

silvertae committed:
Merge branch 'dev' into feat/#64-ItemDetails

silvertae committed:
design #64: 상품상세 페이지 상단바, 하단바 UI 작업

silvertae committed:
design #64: 상세 페이지 본문 UI 구현

silvertae committed:
feat #64: 상품상세 페이지에서 사용할 ImageSlider 컴포넌트 구현

- 브라우저에서 클릭 드래그와 모바일에서 터치 드래그에 작동하도록 구현
- 한 장씩 크기에 맞게 넘어가도록 구현

silvertae committed:
design #64: ImageSlider 내부에 Badge 컴포넌트로 만든 PageNav 추가

silvertae committed:
fix #64: 판매상태 드롭다운 버그 수정

- Status 스타일드 컴포넌트에 position: relative 부여

silvertae committed:
design #64: 드롭다운 관련 스타일 수정

- 홈과 상세 페이지에서 사용할 드롭다운 버튼의 padding 수정

silvertae committed:
feat #64: ProductItem 선택시 해당하는 아이템 상세 페이지 이동 기능 구현

- App에서 상세페이지를 /items/:itemId 라우팅 경로 지정
- ProductItem에 onClick props 추가
- 각 ProductItem 클릭시 해당 아이디 상세페이지로 navigate

silvertae committed:
feat #64: 상세 페이지 mock 데이터를 이용한 렌더링 기능 구현

- 상세 페이지에 사용할 mock data 추가
- ItemDetails 컴포넌트에 useQuery 추가
- MenuItem 선택된 항목 폰트 굵게 표시되는 기능 추가

silvertae committed:
feat #64: 좋아요(하트) 기능 구현

- 관련 fetcher와 mock data 추가
- useMutation 추가
- 좋아요 여부에 따라 Icon 색상 변경 기능 추가

silvertae committed:
feat #64: 상품 판매상태 수정 기능 구현

- 상태 수정 관련 fetcher, mock data 추가
- usePatchStatus mutation 쿼리 구현
- 상태 변경 없을 때 불필요한 요청 방지

silvertae committed:
feat #64: 이미지 슬라이더 페이지 표시 기능 추가

silvertae committed:
feat #64: 구매자 상품 상세 페이지 구현

- data의 isSeller 값에 따라 몇몇 요소 조건부 렌더링
- 상세페이지 Header를 투명하게 변경

silvertae committed:
design #64: Error 컴포넌트 표시를 위해 Footer 위치 조정

silvertae committed:
rename #64: 상세 페이지 관련 api 요청과 쿼리들 분리 및 기존 구조 변경

- itemDetails 관련 axios 요청, 리액트 쿼리, msw 핸들러들을 별도 파일로 분리
- queryKey 별도 파일로 관리
- api 폴더 내 fetch 함수들 fetchers 폴더로 이동

altmit commented on src/api/queries/queryKeys.ts:
```diff
@@ -0,0 +1,8 @@
+export const QUERY_KEY = {
+  ITEMS: 'items',
+  ITEM_DETAILS: 'itemDetails',
+  SALES_LIST: 'salesList',
+  FAVORITES: 'favorites',
+  USER_LOCATION: 'userLocations',
+  LOCATION: 'locations',
+};
```
불필요한 파일 분리라고 생각이 드네요.

이유로는 `ITEMS`는 `useItemQuery`에서만 사용합니다.
`ITEM_DETAILS`는 `useItemDetailsQuery`에서만 사용합니다.

 이미 큰 분류로 파일을 분리했기 때문에 사용처가 한 곳이라서 굳이 파일 분리가 아닌 각각의 파일에 있는게 맞다고 생각합니다.

silvertae replied:
말씀을 듣고 다른 사람들은 queryKey를 어떻게 관리하는지 한 번 조사해봤는데요. 제이 말처럼 query 파일 하나마다 필요한 queryKey를 가까이 두고 사용하는게 좋은 방식이라고 하네요! 유용한 글이 있어 첨부해봅니다. 

이러한 방식을 Colocation 이라고 하나봐요. 아예 필요한 queries.ts 파일을 페이지나 기능 폴더별로 두기도 하는 것 같네요.

https://tkdodo.eu/blog/effective-react-query-keys#colocate

altmit commented on src/components/Footer.tsx:
```diff
@@ -37,6 +37,9 @@ const FooterDiv = styled.div`
   align-items: center;
   padding: 8px 16px;
   border-top: ${({ theme }) => `0.8px solid ${theme.color.neutralBorder}`};
+  position: absolute;
+  bottom: 0;
+  background-color: ${({ theme }) => theme.color.neutralBackgroundWeak};
```
갑자기 어떤 이유로 추가된 css 일까요?

silvertae replied:
기존 isLoading, isError 상태일 때 빈 Layout 화면에서 Footer가 화면 정중앙에 덩그러니 놓여있는 형태였습니다. 빈화면에서도 올바른 위치에 있을 수 있도록 추가하였습니다.

![스크린샷 2023-09-14 오후 5 18 16](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/8c769694-fbd1-4cc8-8a8c-6a86a3190ebe)

이게 너무 싫었네요...

altmit replied:
이렇다면 loading을 주입해 주는 부분에 div같은 element로 감싸고 `flex : 1`을 주는게 맞다고 생각이 드네요.

지금의 css를 그대로 사용한 다면 가정하고  홈에서 Body에 스크롤을 넣는다면 스크롤이 Footer에도 겹쳐저 보여지게 될 것 같네요.

그리고 아마 지금 저희 프로젝트에서는 문제가 없을 확률이 더 높겠지만 이미 너무 많은 곳에서 공용으로 사용하는 컴포넌트를 하나의 특정 조건 때문에 수정하면 어떤 사이드 이팩트가 생길지 모르고 대응하기 힘들 것 같네요

silvertae replied:
음 거기까지는 생각을 못했었네요. 확실히 Home에서 사용하는 Body랑 ItemDetails에서 사용하는 Body랑 조금 다르군요.
일단 원래대로 수정하는게 맞을 것 같아요.

altmit commented on src/components/Header.tsx:
```diff
@@ -3,13 +3,18 @@ import { styled } from 'styled-components';
 type HeaderProps = {
   leftButton?: ReactElement;
-  rightButton?: ReactElement;
+  rightButton?: ReactElement | null;
```
이미 옵셔널로 되어 있는데 null이 추가된 이유는 뭔가요?

silvertae replied:
렌더링을 안 할 때는 null을 넘겨주려고 하다보니까 타입에러가 나서 추가한건데 그냥 undefined를 넘겨주면 해결되는 문제였네요.

altmit commented on src/components/itemDetails/ImageSlider.tsx:
```diff
@@ -0,0 +1,124 @@
+    autoSlide();
+  };
+
+  const autoSlide = () => {
```
https://web.dev/css-scroll-snap/
https://snap.glitch.me/carousel.html

남세가 올려주셨던 글인데 자세히는 안 봤지만 적용하면 더 간단할 것 같네요

silvertae replied:
간단히 읽어보긴 했는데 저도 자세히 한번 봐야겠네요 👀

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -0,0 +1,363 @@
+                삭제
+              </MenuItem>
+            </Dropdown>
+          ) : null
```
```tsx
rightButton={
    data.isSeller ? (
      <Dropdown iconName="dots" align="right">
        <MenuItem onAction={fakeAction}>게시글 수정</MenuItem>
        <MenuItem color="systemWarning" onAction={onClickDelete}>
          삭제
        </MenuItem>
      </Dropdown>
    ) : undefined
  }
```

위에서도 언급 했지만 이미 옵셔널이라서 null은 불필요한 type 지정이네요

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -0,0 +1,363 @@
+      <Main>
+        <ImageSlider imageList={data.images} />
+        <Body>
+          <SellorInfo>
```
sellerInfo 오타인가요?

silvertae replied:
앗... 수정하겠습니다

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -0,0 +1,363 @@
+  color: ${({ theme }) => theme.color.neutralTextWeak};
+`;
+
+const ChatCount = styled.span``;
+
+const FavoritesCount = styled.span``;
+
+const ViewCount = styled.span``;
```
아래에 `FooterRight`를 포함해서 style을 안 줄거면 필요 없는 부분인 것 같네요.
그냥 jsx에서 span, div를 직접 사용하는게 좋을 것 같네요

silvertae replied:
이 부분도 수정할까 말까 고민을 했는데 무조건 Styled components로 정의해둘 필요는 없을 것 같네요.
스타일 필요없는 부분은 그냥 element tag로 수정해보겠습니다.

altmit commented on src/components/Header.tsx:
```diff
@@ -3,13 +3,18 @@ import { styled } from 'styled-components';
+  leftButton,
+  rightButton,
+  title,
+  ...rest
```
Badge에 적용된 rest도 포함되는 질문 입니다.

```tsx
type HeaderProps = {
  leftButton?: ReactElement;
  rightButton?: ReactElement | null;
  title?: string;
};
```
이미 Header의 props는 3가지로 정해져 있습니다.
그런데 ...rest를하면 rest에 어떤게 들어오나요?

그리고 사용처에서 어떻게 사용하고 있는지 잘 모르겠네요.
제가 봤을 때는 없이도 잘 작동하고 있는 것 같습니다.

silvertae replied:
PR 설명에 간단히 적어두긴했는데 순수 styled-component가 아닌 일반적인 함수형 컴포넌트를 다른 곳에서 확장시켜 사용할 때 `className`이 props에 없으면 스타일 상속이 정상적으로 적용되지 않아서 그렇습니다. 

사실 { ...rest }가 아니라 `className` 하나만 추가해도 되는데, 그러면 props 타입에 실제로는 사용하지 않는 className을 추가해야 하는게 귀찮아서 그냥 ...rest로 넘겨서 사용하고 있습니다.

참고문헌) https://styled-components.com/docs/basics#styling-any-component

altmit replied:
음 이런 내용이 있군요.
근데 react만 생각하고 이 컴포넌트를 보면 당황 스러울 것 같네요

왜냐하면 이미 props에 type은3개로 고정 지정해 버렸는데 거기서 rest를 받으면 rest는 빈 오브젝트 아닌가?
생각이 들어요.

아마 console.log를 찍어도 빈 obj일 것 같고 rest를 props에 추가함으로`Bagde`  컴포넌트에 className을 줄 수 있는 것도 아니고

어떤 원리로 돌아가는지 모르겠네요?🤔

silvertae replied:
궁금해서 확인해봤는데 `rest`를 콘솔에 찍어보면 className이 나온답니다.

![스크린샷 2023-09-14 오후 6 36 17](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/ddc81f14-1069-40ea-8e49-151c7464c930)

아마 styled-components가 내부 동작원리를 알아야 이해할 수 있지 않을까 싶네요...

altmit replied:
이건 정말 어떻게 사용하는거지 신기하네요  ㅋㅋ
```tsx
<Badge className="test" />
```
실제로 jsx에서 이렇게 하면 에러가 날 것 같은데 말이죠

altmit commented on src/components/ProductItem.tsx:
```diff
@@ -19,6 +19,7 @@ type ItemProps = {
   };
   thumbnailUrl: string;
   isSeller: boolean;
+  onClick: () => void;
```
onClick을 외부에서 받을 필요는 없을 것 같네요.

부모에서 `selectProductItem(item.id)` 이렇게 적용 중 인데 productItem에서도 충분히 가능한 부분이라서 불필요한 props 같습니다.

그리고 onClick이 필수 값으로 되어 있어 적용되지 않은 다른 컴포넌트에서 ts 에러나고 있을 거라 생각이 되네요

silvertae replied:
앗 진짜 판매내역, 관심목록 페이지에서 에러가 나네요. 말씀대로 props로 외부에서 받는 것보다 ProductItem 컴포넌트 내부에서 자체적으로 onClick을 걸어두는게 좋을 것 같네요

silvertae committed:
refactor #64: 리뷰어 피드백 반영

- Header props type에서 불필요한 타입 수정
- ProductItem 자체에서 onClick 함수로 navigate 하도록 수정
- ItemDetails 컴포넌트에서 return 할 때 불필요한 styled component 제거
- MSW 응답코드들 명세서에 맞게 수정

silvertae committed:
refactor #64: queryKey 위치 원상 복구

silvertae committed:
feat #64: 홈화면 동네 설정 드롭다운에서 위치 변경기능 뒤늦은 추가

silvertae committed:
design #64: Footer 컴포넌트 추가된 디자인 삭제
